sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotShiftByZeroOrIntSize.cs

```diff
@@ -1,151 +1,238 @@
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class DoNotShiftByZeroOrIntSize : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S2183";
+
+        private const string MessageFormat_UseLargerTypeOrPromote
+            = "Either promote shift target to a larger integer type or shift by {0} instead.";
+        private const string MessageFormat_ShiftTooLarge = "Correct this shift; shift by {0} instead.";
+        private const string MessageFormat_RightShiftTooLarge = "Correct this shift; '{0}' is larger than the type size.";
+        private const string MessageFormat_UselessShift = "Remove this useless shift by {0}.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, "{0}", RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private static Dictionary<KnownType, int> mapKnownTypesToIntegerBitSize
+            = new Dictionary<KnownType, int>
+            {
+                [KnownType.System_Int64] = 64,
+                [KnownType.System_UInt64] = 64,
+
+                [KnownType.System_Int32] = 32,
+                [KnownType.System_UInt32] = 32,
+
+                [KnownType.System_Int16] = 16,
+                [KnownType.System_UInt16] = 16,
+
+                [KnownType.System_Byte] = 8,
+                [KnownType.System_SByte] = 8
+            };
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var shiftInstances = (c.Node as MemberDeclarationSyntax)
+                        .DescendantNodes()
+                        .Select(n => FindIssues(n, c.SemanticModel))
+                        .WhereNotNull();
+
+                    var zeroShiftIssues = new List<ShiftInstance>();
+                    var linesWithShiftOperations = new HashSet<int>();
+
+                    foreach (var shift in shiftInstances)
+                    {
+                        if (shift.IsLiteralZero)
+                        {
+                            zeroShiftIssues.Add(shift);
+                            continue;
+                        }
+
+                        linesWithShiftOperations.Add(shift.Line);
+                        if (shift.Diagnostic != null)
+                        {
+                            c.ReportDiagnosticWhenActive(shift.Diagnostic);
+                        }
+                    }
+
+                    foreach (var issue in zeroShiftIssues)
+                    {
+                        if (!linesWithShiftOperations.Contains(issue.Line - 2) &&

```

Comment:
This list of conditions could be refactored either in a method or using a collection of lines and a diff. But definitely this looks bad when reading.